lib/streamlit/connections/sql_connection.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import TYPE_CHECKING, Iterator, List, Optional, Union, cast
+
+import pandas as pd
+
+from streamlit.connections import ExperimentalBaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine.base import Engine
+    from sqlalchemy.orm import Session
+
+
+_REQUIRED_CONNECTION_PARAMS = {"dialect", "username", "host"}
+
+
+class SQL(ExperimentalBaseConnection["Engine"]):
+    def _connect(self, autocommit: bool = False, **kwargs) -> "Engine":
+        import sqlalchemy
+
+        secrets = self._secrets
+
+        if "url" in secrets:
+            url = sqlalchemy.engine.make_url(secrets["url"])
+        else:
+            for p in _REQUIRED_CONNECTION_PARAMS:
+                if p not in secrets:
+                    raise StreamlitAPIException(f"Missing SQL DB connection param: {p}")
+
+            drivername = secrets["dialect"] + (
+                f"+{secrets['driver']}" if "driver" in secrets else ""
+            )
+
+            url = sqlalchemy.engine.URL.create(
+                drivername=drivername,
+                username=secrets["username"],
+                password=secrets.get("password"),
+                host=secrets["host"],
+                port=int(secrets["port"]) if "port" in secrets else None,
+                database=secrets.get("database"),
+            )
+
+        eng = sqlalchemy.create_engine(url, **kwargs)
```

Comment:
_(Duplicating my comment from https://github.com/streamlit/streamlit/pull/6486 in case it isn't addressed there)_

snowflake-sqlalchemy uses a connect_args kwarg outside the URL to specify common behaviors like non-password based authentication ([rough cite](https://docs.snowflake.com/developer-guide/python-connector/sqlalchemy#key-pair-authentication-support) - this is also what I found when I tried it out).

Similarly, bigquery-sqlalchemy uses a credentials_info kwarg which seems to be required for any usage ([cite](https://github.com/googleapis/python-bigquery-sqlalchemy#authentication)).

It seems like there's no way to get that passed into create_engine directly from secrets with this implementation. So to use these dialects (which are pretty important!) I still have to do something like:
conn = st.connection('sql', connect_args=dict(st.secrets.snowflakeargs))

I would love to be able to just specify a secrets section like [connections.sql.connect_args] and have it get passed through.

Reply:
I think this name will have to be something a bit clunkier like `[connections.sql.create_engine_kwargs.connect_args]` since I think it'll be unreasonable to ensure that every possible dialect-specific create_engine kwarg can make it here, but aside from that this should be pretty straightforward.

Will get a PR out for this momentarily